### PR TITLE
Poisson zero rate

### LINKED ISSUE #53485: Poisson function failed on rate vector contain zeros.
## 🐛 Bug

Poisson function failed on rate vector contain zeros. 

## To Reproduce

Run the following code on 
```
r = torch.tensor([0, 0.1, 0.2, 0.3])
p = torch.distributions.poisson.Poisson(rate=r)
```
PyTorch 1.7.1  run fine
PyTorch 1.8  getting ``Exception has occurred: ValueError
The parameter rate has invalid values``

## Expected behavior

Zeros should be acceptable value

## Environment
```
PyTorch version: 1.8.0+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3

Python version: 3.8 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.1
[pip3] torch==1.8.0+cpu
[pip3] torchaudio==0.8.0
[pip3] torchvision==0.9.0+cpu
[conda] Could not collect
```
OR 
```
PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 2070
Nvidia driver version: 460.39
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.4
[pip3] torch==1.7.1
[pip3] torchvision==0.8.2
[conda] Could not collect
```

cc @fritzo @neerajprad @alicanb @vishwakftw @nikitaved

### PR DESCRIPTION
**Pull request by tillahoffmann**
This PR fixes #53485 by allowing zero rates for the Poisson distribution. This implementation is consistent with `scipy.stats.poisson` which admits zero rates. In addition to addressing the aforementioned issue, this PR makes two supporting changes:

1. add a `nonnegative` constraint to enforce non-negative rates for the Poisson distribution.
2. adjust the evaluation of the gradient of `xlogy` such that it is well defined for `x == 0 and y == 0`.

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -786,11 +786,11 @@
 
 - name: xlogy.Tensor(Tensor self, Tensor other) -> Tensor
   self: grad * at::xlogy((self != 0), other)
-  other: grad * self / other
+  other: grad * at::where(other.isnan() | (self != 0), self / other, zeros_like(other))
```
Comment (fritzo) [inline tools/autograd/derivatives.yaml:789]: Can someone confirm that this does not suffer from #17455?

File: torch/distributions/constraints.py [hunk 2]
```diff
@@ -545,6 +545,7 @@ def check(self, value):
 real = _Real()
 real_vector = independent(real, 1)
 positive = _GreaterThan(0.)
+nonnegative = _GreaterThanEq(0.)
```
Comment (fritzo) [inline torch/distributions/constraints.py:548]: Would you mind also registering this like `positive` in the constraint registry, so as to avoid breaking downstream code that uses the constraint registry to register constrained parameters, e.g. when learning parameters of Poisson distributions?
```diff
diff --git a/torch/distributions/constraint_registry.py b/torch/distributions/constraint_registry.py
index cbe987e72c..c03f0ad02d 100644
--- a/torch/distributions/constraint_registry.py
+++ b/torch/distributions/constraint_registry.py
@@ -173,7 +173,9 @@ def _transform_to_independent(constraint):


 @biject_to.register(constraints.positive)
+@biject_to.register(constraints.nonnegative)
 @transform_to.register(constraints.positive)
+@transform_to.register(constraints.nonnegative)
 def _transform_to_positive(constraint):
     return transforms.ExpTransform()
```

File: torch/distributions/poisson.py [hunk 3]
```diff
@@ -60,7 +60,7 @@ def log_prob(self, value):
         if self._validate_args:
             self._validate_sample(value)
         rate, value = broadcast_all(self.rate, value)
-        return (rate.log() * value) - rate - (value + 1).lgamma()
+        return value.xlogy(rate) - rate - (value + 1).lgamma()
```
Comment (fritzo) [inline torch/distributions/poisson.py:63]: Nice :+1:

**Review by fritzo — COMMENTED** [review]
Implementation and tests look good to me.  My only concerns are the constraint registries and interaction with #17455 (though the latter should not break backwards compatibility).

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (tillahoffmann) [reply tools/autograd/derivatives.yaml:789]: I can't comment on the specifics of #17455, but the change introduced here is in part to mitigate the presence of `nan`s where they are not expected in the backwards pass. Consider the following example.

```python
import torch as th

for x in [0., 2., float('nan')]:
    for y in [0., 1.3, float('nan')]:
        x = th.as_tensor(x).requires_grad_()
        y = th.as_tensor(y).requires_grad_()
        th.xlogy(x, y).backward()
        print(f'df/dx({x:.3f}, {y:.3f}) = {x.grad:.3f}')
        print(f'df/dy({x:.3f}, {y:.3f}) = {y.grad:.3f}')
```

Running this script on `master` and this branch and comparing the outputs yields.

```diff
diff --git a/old.txt b/new.txt
index 1141aef386..0fe45e40ed 100644
--- a/old.txt
+++ b/new.txt
@@ -1,18 +1,18 @@
 df/dx(0.000, 0.000) = 0.000
-df/dy(0.000, 0.000) = nan
+df/dy(0.000, 0.000) = 0.000
 df/dx(0.000, 1.300) = 0.000
 df/dy(0.000, 1.300) = 0.000
 df/dx(0.000, nan) = nan
 df/dy(0.000, nan) = nan
 df/dx(2.000, 0.000) = -inf
 df/dy(2.000, 0.000) = inf
 df/dx(2.000, 1.300) = -inf
 df/dy(2.000, 1.300) = 1.538
 df/dx(2.000, nan) = nan
 df/dy(2.000, nan) = nan
 df/dx(nan, 0.000) = -inf
 df/dy(nan, 0.000) = nan
 df/dx(nan, 1.300) = -inf
 df/dy(nan, 1.300) = nan
 df/dx(nan, nan) = nan
 df/dy(nan, nan) = nan
```

In other words, the presence of `nan`s in gradients should only be reduced as a result of merging this PR.

**Comment by tillahoffmann** [comment]
Hi @fritzo, do you require any additional changes to this PR?

**Review by fritzo — APPROVED** [review]
LGTM

**Comment by tillahoffmann** [comment]
Hi @albanD, do you have any additional changes as code owner or is this one ready to be merged?

**Review by albanD — APPROVED** [review]
LGTM

**Comment by albanD** [comment]
Can you rebase on top of master and make sure all tests are good please?

**Comment by tillahoffmann** [comment]
The branch is now rebased on `master`. A few tests are failing, but it looks like they're related to setting up the environment/hitting a time limit/running out of resources rather than the changes in this PR.

**Comment by albanD** [comment]
Yes CI looks flaky.
Thanks for the rebase!
